File: wpimath/src/test/native/cpp/estimator/MerweScaledSigmaPointsTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/math/estimator/MerweScaledSigmaPoints.hpp"

#include <gtest/gtest.h>

TEST(MerweScaledSigmaPointsTest, ZeroMean) {
  wpi::math::MerweScaledSigmaPoints<2> sigmaPoints;
  auto points = sigmaPoints.SquareRootSigmaPoints(
      wpi::math::Vectord<2>{0.0, 0.0},
      wpi::math::Matrixd<2, 2>{{1.0, 0.0}, {0.0, 1.0}});

  EXPECT_TRUE(
      (points -
       wpi::math::Matrixd<2, 5>{{0.0, 0.00173205, 0.0, -0.00173205, 0.0},
                                {0.0, 0.0, 0.00173205, 0.0, -0.00173205}})
          .norm() < 1e-3);
}

TEST(MerweScaledSigmaPointsTest, NonzeroMean) {
  wpi::math::MerweScaledSigmaPoints<2> sigmaPoints;
  auto points = sigmaPoints.SquareRootSigmaPoints(
      wpi::math::Vectord<2>{1.0, 2.0},
      wpi::math::Matrixd<2, 2>{{1.0, 0.0}, {0.0, std::sqrt(10.0)}});

  EXPECT_TRUE(
      (points - wpi::math::Matrixd<2, 5>{{1.0, 1.00173205, 1.0, 0.998268, 1.0},
                                         {2.0, 2.0, 2.00548, 2.0, 1.99452}})
          .norm() < 1e-3);
}

File: wpimath/src/test/java/edu/wpi/first/math/estimator/MerweScaledSigmaPointsTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.math.estimator;

import static org.junit.jupiter.api.Assertions.assertTrue;

import edu.wpi.first.math.Matrix;
import edu.wpi.first.math.Nat;
import edu.wpi.first.math.VecBuilder;
import org.junit.jupiter.api.Test;

class MerweScaledSigmaPointsTest {
  @Test
  void testZeroMeanPoints() {
    var merweScaledSigmaPoints = new MerweScaledSigmaPoints<>(Nat.N2());
    var points =
        merweScaledSigmaPoints.squareRootSigmaPoints(
            VecBuilder.fill(0, 0), Matrix.mat(Nat.N2(), Nat.N2()).fill(1, 0, 0, 1));

    assertTrue(
        points.isEqual(
            Matrix.mat(Nat.N2(), Nat.N5())
                .fill(
                    0.0, 0.00173205, 0.0, -0.00173205, 0.0, 0.0, 0.0, 0.00173205, 0.0, -0.00173205),
            1E-6));
  }

  @Test
  void testNonzeroMeanPoints() {
    var merweScaledSigmaPoints = new MerweScaledSigmaPoints<>(Nat.N2());
    var points =
        merweScaledSigmaPoints.squareRootSigmaPoints(
            VecBuilder.fill(1, 2), Matrix.mat(Nat.N2(), Nat.N2()).fill(1, 0, 0, Math.sqrt(10)));

    assertTrue(
        points.isEqual(
            Matrix.mat(Nat.N2(), Nat.N5())
                .fill(1.0, 1.00173205, 1.0, 0.99826795, 1.0, 2.0, 2.0, 2.00547723, 2.0, 1.99452277),
            1E-6));
  }
}
